src/storage/secondary/block/dict_block_builder.rs

```diff
@@ -0,0 +1,249 @@
+// Copyright 2022 RisingLight Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use bytes::BufMut;
+use risinglight_proto::rowset::BlockStatistics;
+
+use super::PlainPrimitiveBlockBuilder;
+use crate::array::{Array, I32Array};
+use crate::storage::secondary::block::{BlockBuilder, RleBlockBuilder};
+
+pub(crate) const DICT_NULL_VALUE_KEY: i32 = i32::MIN;
+/// Encodes fixed-width data into a block with dict encoding. The layout is
+/// ```plain
+/// | rle_block_length(u64) |  dict_count_sum (u32) | rle_block | dict_block |
+/// ```
+pub struct DictBlockBuilder<A, B>
+where
+    A: Array,
+    B: BlockBuilder<A>,
+    A::Item: PartialEq,
+    <A::Item as ToOwned>::Owned: Eq + Hash,
+{
+    dict_map: HashMap<<A::Item as ToOwned>::Owned, i32>,
+    data_builder: B,
+    rle_builder: RleBlockBuilder<I32Array, PlainPrimitiveBlockBuilder<i32>>,
```

Comment:
Spent some time to understand why `rle_builder` is correct here  🤣 I'd recommend you to add some docstring here so that other developers can understand this easily.